### ISSUE TITLE
Improving CI for new packages and updates

### ISSUE THREAD

**Issue by bkotzz** [cpp-pm/hunter#9]
The current state of the world for CI in Hunter is as follows:
- PRs against the proper hunter repo only run CI related to the docs
- Per-package CI exists, but is complicated. The instructions are detailed [here](https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#branch-pkg-template), but the general idea is that there is a [fork of hunter](https://github.com/ingenue/hunter) that has a branch per-package, and the only relevant files on each branch are the Travis/Appveyor files, to be able to run tests for that specific package on that branch
- It is acceptable for packages to not build on certain platforms and still be accepted. The relevant builds in the build matrix for that package just end up getting commented out on the testing branch
- Someone has to go and create the testing branch for new packages before they can run tests
- For both new and existing packages, PRs against hunter proper must include links to passing Travis/Appveyor builds in the description, on a branch that has the testing branch _merged into the feature branch_, such that it runs the package's tests against the feature's changes. Someone then has to go and look at the builds to make sure the setup is correct, and that they passed

A new approach that should streamline how we test packages:
- Every package has a testing table, that specifies which builds that package is blacklisted for. The set of all possible builds comes from the [Travis](https://github.com/ingenue/hunter/blob/pkg.template/.travis.yml) and [AppVeyor](https://github.com/ingenue/hunter/blob/pkg.template/appveyor.yml) testing table. A package's blacklist can be empty, indicating it passes all builds. This blacklist can possibly be a YML file in the `cmake/projects/` folder
- For any pull request against hunter, in CI we identify the changed files in the PR. For any changes to `cmake/projects/` or `examples/`, the package(s) being changed is noted. The list of packages being changed in a given PR can be empty (ie. changes to core Hunter only), but will generally be length 1.
- For every Travis/AppVeyor build in the build matrix, for every package in the changed packages list, we build the package on that platform. If no packages are changed, we exit immediately.
- To prevent builds that last for hours, we can possibly limit how many packages can be changed in a given PR - such that changes to large sets of packages need to be split into smaller PRs. However, the build times per package per platform are generally quite fast.

Summary:
We would no longer have to use `ingenue/hunter`, and instead would run CI for the packages being changed directly on the PR against hunter proper

**Comment by NeroBurner** [comment]
how and where are the changes in `cmake/projects` or `examples` detected? How do you decide which versions to compare?

**Comment by bkotzz** [comment]
How - using a git diff (name only) between the HEAD commit of the PR branch, and the merge base of the PR branch and the merge target (which is the master branch of this repo). Essentially the detected changes files should be identical to the changed files you see in the PR on github

Where - at the beginning of each build in the build matrix for Travis and appveyor jobs

Which versions - I think I answered that above, but let me know if something doesn’t make sense

**Comment by bkotzz** [comment]
A proposal for an intermediate step:

We have currently moved `ingenue/hunter` into the cpp-pm org here https://github.com/cpp-pm/hunter-testing. Again, this repo's only relation to hunter is that each branch contains the travis/appveyor build matrix for that package. It doesn't need to be a fork of hunter, because instead of merging the testing branches into feature branches, it's reasonable to do:
```
# On your feature branch

git remote add testing https://github.com/cpp-pm/hunter-testing
git fetch testing pkg.gtest
git checkout testing/pkg.gtest -- appveyor.yml
git checkout testing/pkg.gtest -- .travis.yml
git checkout -b gtest_test
git commit -m "Test gtest builds"
git push -u origin gtest_test

# As long as CI is enabled on your fork, go to travis/appveyor websites and track the relevant build job that you just kicked off
```

So instead of maintaining a repo with branches to hold the build matrices for each package, we could instead check them into https://github.com/cpp-pm/hunter as unused files. So for example, we would have `cmake/projects/GTest/.travis.yml` and `cmake/projects/GTest/appveyor.yml`. Then, if you wanted to run CI when updating gtest to a new version, you would do:
```
cp cmake/projects/GTest/.travis.yml .
cp cmake/projects/GTest/appveyor.yml .
git checkout -b gtest_test
git commit -m "Test gtest builds"
git push -u origin gtest_test
```

This would then mean that:
- To add a new package, you don't need a PR against the testing repo to create the branch for the new package
- For existing packages, you don't need to merge against another repo to run CI
- When updating packages, if a build in the build matrix starts/stops failing, you can comment/uncomment in the same PR against hunter that updates the package.

**Comment by NeroBurner** [comment]
GitLab released generated child pipelines. You can generate the jobs, that need to be triggered dynamically. So one job detects which examples need to be run, and a child pipeline then runs only for the changed project files
https://docs.gitlab.com/ee/ci/parent_child_pipelines.html#dynamic-child-pipelines

is there something like that in the github space?

**Comment by rbsheth** [comment]
Closed by #250
